core/lib/dal/src/connection/processor.rs

```diff
@@ -0,0 +1,222 @@
+
+/// Tags that can be associated with a connection.
+#[derive(Debug, Clone, Copy)]
+pub(super) struct StorageProcessorTags {
+    pub requester: &'static str,
+    pub location: &'static Location<'static>,
+}
+
+impl fmt::Display for StorageProcessorTags {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            formatter,
+            "requested by `{}` at {}:{}",
+            self.requester,
+            self.location.file(),
+            self.location.line()
+        )
+    }
+}
+
+struct TracedConnectionInfo {
+    tags: Option<StorageProcessorTags>,
+    created_at: Instant,
+}
+
+impl fmt::Debug for TracedConnectionInfo {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let timestamp: chrono::DateTime<chrono::Utc> =
+            (SystemTime::now() - self.created_at.elapsed()).into();
+        let tags_display: &dyn fmt::Display = match &self.tags {
+            Some(tags) => tags,
+            None => &"not tagged",
+        };
+        write!(formatter, "[{timestamp} - {tags_display}]")
+    }
+}
+
+/// Traced active connections for a connection pool.
+#[derive(Default)]
+pub(super) struct TracedConnections {
+    connections: Mutex<HashMap<usize, TracedConnectionInfo>>,
+    next_id: AtomicUsize,
+}
+
+impl fmt::Debug for TracedConnections {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Ok(guard) = self.connections.lock() {
+            formatter.debug_set().entries(guard.values()).finish()
+        } else {
+            formatter.write_str("(poisoned)")
+        }
+    }
+}
+
+impl TracedConnections {
+    fn acquire(&self, tags: Option<StorageProcessorTags>, created_at: Instant) -> usize {
+        let id = self.next_id.fetch_add(1, Ordering::SeqCst);
+        let mut guard = self
+            .connections
+            .lock()
+            .expect("`TracedConnections` is poisoned");
+        let info = TracedConnectionInfo { tags, created_at };
+        guard.insert(id, info);
+        id
+    }
+
+    fn mark_as_dropped(&self, connection_id: usize) {
+        let mut guard = self
+            .connections
+            .lock()
+            .expect("`TracedConnections` is poisoned");
+        guard.remove(&connection_id);
+    }
+}
+
+struct PooledStorageProcessor<'a> {
+    connection: PoolConnection<Postgres>,
+    tags: Option<StorageProcessorTags>,
+    created_at: Instant,
+    traced: Option<(&'a TracedConnections, usize)>,
+}
+
+impl fmt::Debug for PooledStorageProcessor<'_> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("PooledStorageProcessor")
+            .field("tags", &self.tags)
+            .field("created_at", &self.created_at)
+            .finish_non_exhaustive()
+    }
+}
+
+impl Drop for PooledStorageProcessor<'_> {
+    fn drop(&mut self) {
+        const LONG_CONNECTION_THRESHOLD: Duration = Duration::from_secs(5);
```

Comment:
Not sure what the reasonable value for this param should be.

Reply:
I cannot wait to see this data!!!

As for the threshold, I think something between 2 and 5 is good.